File: typescript/src/random.ts
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * TypeScript rewrite derived from NPGenerator V2.0.2.
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba / Programmer: Masaya Kiwada
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

const MULTIPLIER = 0x5deece66dn;
const ADDEND = 0xbn;
const MASK = (1n << 48n) - 1n;

/** java.util.Random-compatible 48-bit LCG. */
export class JavaRandom {
  private state = 0n;

  constructor(seed: bigint | number) {
    this.setSeed(seed);
  }

  setSeed(seed: bigint | number): void {
    this.state = (BigInt(seed) ^ MULTIPLIER) & MASK;
  }

  next(bits: number): number {
    if (bits < 0 || bits > 32) throw new RangeError("bits must be between 0 and 32");
    this.state = (this.state * MULTIPLIER + ADDEND) & MASK;
    return Number(this.state >> BigInt(48 - bits));
  }

  nextInt(bound: number): number {
    if (!Number.isInteger(bound) || bound <= 0 || bound > 0x7fffffff) {
      throw new RangeError("bound must be positive");
    }
    if ((bound & -bound) === bound) {
      return Number((BigInt(bound) * BigInt(this.next(31))) >> 31n);
    }
    let bits: number;
    let value: number;
    do {
      bits = this.next(31);
      value = bits % bound;
      // Reproduce signed Java int overflow in the rejection test.
    } while (((bits - value + (bound - 1)) | 0) < 0);
    return value;
  }

  shuffle<T>(values: T[]): void {
    for (let i = values.length; i > 1; i--) {
      const j = this.nextInt(i);
      [values[i - 1], values[j]] = [values[j]!, values[i - 1]!];
    }
  }
}

File: rust/src/random.rs
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * Rust rewrite derived from NPGenerator V2.0.2.
 *
 * Number Place Generator Version 2.0
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba
 * Programmer: Masaya Kiwada
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

pub struct JavaRandom {
    state: u64,
}

impl JavaRandom {
    const MULTIPLIER: u64 = 0x5DEECE66D;
    const ADDEND: u64 = 0xB;
    const MASK: u64 = (1_u64 << 48) - 1;

    pub fn new(seed: i64) -> Self {
        Self {
            state: ((seed as u64) ^ Self::MULTIPLIER) & Self::MASK,
        }
    }

    fn next(&mut self, bits: u32) -> i32 {
        self.state = self
            .state
            .wrapping_mul(Self::MULTIPLIER)
            .wrapping_add(Self::ADDEND)
            & Self::MASK;
        (self.state >> (48 - bits)) as i32
    }

    pub fn next_int(&mut self, bound: i32) -> i32 {
        assert!(bound > 0);
        if (bound & -bound) == bound {
            return ((bound as i64 * self.next(31) as i64) >> 31) as i32;
        }
        loop {
            let bits = self.next(31);
            let value = bits % bound;
            if bits.wrapping_sub(value).wrapping_add(bound - 1) >= 0 {
                return value;
            }
        }
    }

    pub fn shuffle<T>(&mut self, values: &mut [T]) {
        for i in (2..=values.len()).rev() {
            let j = self.next_int(i as i32) as usize;
            values.swap(i - 1, j);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::JavaRandom;

    #[test]
    fn java_sequence() {
        let mut random = JavaRandom::new(1);
        assert_eq!(random.next_int(9), 6);
        assert_eq!(random.next_int(9), 1);
        assert_eq!(random.next_int(9), 1);
    }
}

